src/sql-pretty/src/doc.rs

```diff
@@ -17,26 +17,39 @@ use crate::util::{
     bracket, bracket_doc, comma_separate, comma_separated, intersperse_line_nest, nest,
     nest_comma_separate, nest_title, title_comma_separate,
 };
-use crate::TAB;
+use crate::{PrettyConfig, TAB};
 
 // Use when we don't know what to do.
-pub(crate) fn doc_display<'a, T: AstDisplay>(v: &T, _debug: &str) -> RcDoc<'a, ()> {
+pub(crate) fn doc_display<'a, T: AstDisplay>(
+    v: &T,
+    config: PrettyConfig,
+    _debug: &str,
+) -> RcDoc<'a, ()> {
     #[cfg(test)]
     eprintln!(
         "UNKNOWN PRETTY TYPE in {}: {}, {}",
         _debug,
         std::any::type_name::<T>(),
-        v.to_ast_string()
+        v.to_ast_string_simple()
     );
-    doc_display_pass(v)
+    doc_display_pass(v, config)
 }
 
 // Use when the AstDisplay trait is what we want.
-fn doc_display_pass<'a, T: AstDisplay>(v: &T) -> RcDoc<'a, ()> {
-    RcDoc::text(v.to_ast_string())
+fn doc_display_pass<'a, T: AstDisplay>(v: &T, config: PrettyConfig) -> RcDoc<'a, ()> {
+    RcDoc::text(v.to_ast_string(config.format_mode))
 }
 
-pub(crate) fn doc_create_source<T: AstInfo>(v: &CreateSourceStatement<T>) -> RcDoc {
+fn doc_display_pass_mapper<T: AstDisplay>(
+    config: PrettyConfig,
+) -> impl for<'b> FnMut(&'b T) -> RcDoc<'b, ()> {
+    move |v| doc_display_pass(v, config)
+}
+
+pub(crate) fn doc_create_source<T: AstInfo>(
+    v: &CreateSourceStatement<T>,
+    config: PrettyConfig,
+) -> RcDoc {
```

Comment:
It seems the PR needs to introduce a bunch of complexity to work around the existing code structure. It strikes me as potentially the wrong approach: Why don't we convert the freestanding functions to functions on a type instead? That way we wouldn't need to pass the config everywhere.

Reply:
Yeah, I'll probably do this. I don't think it will really reduce the complexity, because the `self` parameter will just take the place of the current `config` parameter, but should help readability a bit.

(But first, I'd like to resolve [the question of whether we even need to plumb `FormatMode` through the `sql-pretty` crate](https://github.com/MaterializeInc/materialize/pull/31933#discussion_r2017283330). If not, then most of the changes to `sql-pretty` can simply be reverted.)